### PR TITLE
Add io feature to OTF

### PR DESCRIPTION
**Pull request by YuuuXie**
## Summary

* Feature: add `write_model=4` in otf. This mode will back up all checkpoint files after each DFT call. For example, the DFT calls are made at step 1, 11, 23, ..., then there will be folders `ckpt_1`, `ckpt_11`, `ckpt_23` created that in each folder the checkpoint files at that step are stored.

In this way, the OTF can restart from any of the previous DFT steps easily with the files and `OTF.from_checkpoint`. This is helpful when the OTF training sometimes gets unphysical after some point in the simulation, e.g. a certain DFT fails to converge but the MD still goes on with the unconverged data, or the gp hyps optimization fails to converge, etc.

* Other updates in otf output:
  * recording package version numbers in the header of output file
  * recording per-species MAEs after each DFT call
  * recording DFT data into an .xyz file including sparse environments (target_atoms indices). This .xyz file will include all and only the training frames (and their corresponding sparse environment indices). After the training is done (or when we want to build a master potential from multiple OTFs), it might be easier to build SGP from this file instead of parsing the otf trajectory. Also if we want to release a training data set or give training data to other people (e.g. the NN folks), we can use this file or combine multiple .xyz files from different OTFs.
  * updating the OTF parser

## Checklist

Before a pull request can be merged, the following items must be checked:

- [x] Code is in the [standard Python style](https://www.python.org/dev/peps/pep-0008/). 
      Run [Black](https://pypi.org/project/black/) on your local machine.
- [x] Docstrings have been added in the [Sphinx docstring format](https://sphinx-rtd-tutorial.readthedocs.io/en/latest/docstrings.html).
- [x] Type annotations are **highly** encouraged.
- [x] Tests have been added for any new functionality or bug fixes.
- [x] All existing tests pass.
- [x] The version number in `setup.py` is updated. We are using a version number format a.b.c
    - If this PR fixes bugs, update version number to a.b.c+1
    - If this PR adds new features, update version number to a.b+1.0
    - If this PR includes significant changes in framework or interface, update version number to a+1.0.0

Note that the CI system will run all the above checks. But it will be much more
efficient if you already fix most errors prior to submitting the PR.

### CODE REVIEW

File: flare/ase/otf.py [hunk 1]
```diff
@@ -325,15 +362,28 @@ def as_dict(self):
...
+            flare_calc = FLARE_Calculator.from_file(dct["flare_calc"])
+            _kernels = None
+        elif flare_calc_dict["class"] == "SGP_Calculator":
+            from flare_pp.sparse_gp_calculator import SGP_Calculator
+
+            flare_calc, _kernels = SGP_Calculator.from_file(dct["flare_calc"])
+        else:
+            raise TypeError(
+                f"The calculator from {dct['flare_calc']} is not recognized."
+            )
```
Comment (jonpvandermause) [inline flare/ase/otf.py:382]: These changes are a bit hard to follow -- a comment would be helpful

File: flare/ase/otf.py [hunk 2]
```diff
@@ -260,8 +272,26 @@ def update_gp(self, train_atoms, dft_frcs, dft_energy=None, dft_stress=None):
             flare_stress = None
 
         # update gp model
+        try:
```
Comment (jonpvandermause) [inline flare/ase/otf.py:275]: A comment justifying this try/except block would be helpful (e.g. why is a deepcopy needed here?)

File: flare/otf.py [hunk 3]
```diff
@@ -467,13 +485,55 @@ def train_gp(self):
...
+            )
+            per_species_mav[species_ind] += np.mean(np.abs(dft_forces[a]))
+            per_species_num[species_ind] += 1
+        per_species_mae /= per_species_num
+        per_species_mav /= per_species_num
+
+        for s in range(len(unique_species)):
+            curr_species = unique_species[s]
+            f.info(f"type {curr_species} forces mae: {per_species_mae[s]:.4f} eV/A")
+            f.info(f"type {curr_species} forces mav: {per_species_mav[s]:.4f} eV/A")
```
Comment (jonpvandermause) [inline flare/otf.py:536]: Consider including this change in the PR description? Will help to keep track of what the changes are

File: flare/otf.py [hunk 4]
```diff
@@ -514,6 +574,18 @@ def record_state(self):
...
+        self.output.write_xyz_config(
+            self.curr_step,
+            self.structure,
+            self.structure.forces,
+            self.structure.stds,
+            dft_forces=self.structure.forces,
+            dft_energy=self.structure.potential_energy,
+            target_atoms=target_atoms,
+        )
+
```
Comment (jonpvandermause) [inline flare/otf.py:588]: Should be helpful for keeping track of the sparse environments. I would include this in the PR description too

File: flare/otf.py [hunk 5]
```diff
@@ -265,27 +271,36 @@ def run(self):
...
-                    self.steps_since_dft = 0
+                    self.last_dft_step = self.curr_step
                     self.run_dft()
                     dft_frcs = deepcopy(self.structure.forces)
                     dft_stress = deepcopy(self.structure.stress)
                     dft_energy = self.structure.potential_energy
 
                     # run MD step & record the state
                     self.record_state()
+                    self.record_dft_data(self.structure, target_atoms)
```
Comment (jonpvandermause) [inline flare/otf.py:293]: A more detailed comment here would be helpful -- e.g. where is the xyz file being stored?

File: flare/output.py [hunk 6]
```diff
@@ -134,6 +136,55 @@ def write_header(
         f = logging.getLogger(self.basename + "log")
         f.info(f"{datetime.datetime.now()}")
 
+        # Write package version numbers
```
Comment (jonpvandermause) [inline flare/output.py:139]: Good idea. Can you add this to the PR description?

**Review by jonpvandermause — COMMENTED** [review]
Looks good to me, main comment is to make sure the PR description gives a complete summary of the changes (e.g. recording per-species MAEs and version numbers in the output file, recording DFT data and sparse environments, updating the OTF parser, etc.)

**Comment by YuuuXie** [comment]
> Looks good to me, main comment is to make sure the PR description gives a complete summary of the changes (e.g. recording per-species MAEs and version numbers in the output file, recording DFT data and sparse environments, updating the OTF parser, etc.)

Oh yes absolutely, thanks for reviewing and pointing this out! I'll go ahead merge it.

File: flare/otf.py (hunk 3 above)
Comment (YuuuXie) [reply flare/otf.py:536]: added

File: flare/otf.py (hunk 4 above)
Comment (YuuuXie) [reply flare/otf.py:588]: added to the description

File: flare/output.py (hunk 6 above)
Comment (YuuuXie) [reply flare/output.py:139]: added

File: flare/otf.py (hunk 5 above)
Comment (YuuuXie) [reply flare/otf.py:297]: added in commit f3c0b55

File: flare/ase/otf.py (hunk 2 above)
Comment (YuuuXie) [reply flare/ase/otf.py:277]: added in commit f3c0b55

File: flare/ase/otf.py (hunk 1 above)
Comment (YuuuXie) [reply flare/ase/otf.py:394]: the kernel needs to be returned, especially when we build SGP. It is still the issue we had before that if the kernel is not returned to the current space, the variable of it (like `kernel->hyperparameters`) can not be found and raises segmentation fault.

comments are added in commit f3c0b55

**Comment by jonpvandermause** [comment]
> > Looks good to me, main comment is to make sure the PR description gives a complete summary of the changes (e.g. recording per-species MAEs and version numbers in the output file, recording DFT data and sparse environments, updating the OTF parser, etc.)
> 
> Oh yes absolutely, thanks for reviewing and pointing this out! The descriptions are updated

Looks good! Feel free to merge
